### PR TITLE
refactor: use environment param instead of context.environment

### PR DESCRIPTION
**Pull request by 9aoy**
## Summary

remove `api.context.environments`, use `environment` param instead.
- add `environment` param into  `modifyBundlerChain` / `modifyRspackConfig` / `modifyHTMLTags` hooks
- add `environments` param into  `onAfterBuild` / `onAfterCreateCompiler`  / `onBeforeStartDevServer`   / `onBeforeBuild`  / `onAfterBuild` hooks

```ts 
type EnvironmentContext = {
  index: number;
  name: string;
  entry: RsbuildEntry;
  htmlPaths: Record<string, string>;
  distPath: string;
  browserslist: string[];
  tsconfigPath?: string;
  config: NormalizedEnvironmentConfig;
};


type ModifyBundlerChainUtils = {
  environment: string;
  environment: EnvironmentContext;
};

```

```diff
const myPlugin = () => ({
  setup: (api) => {
    api.modifyBundlerChain((chain, utils) => {
-      const { distPath } = api.context.environments[utils.environment];
+      const { distPath } = utils.environment;

-      const environmentConfig = api.getNormalizedConfig({ environment });
+      const environmentConfig = environment.config;
    });
  },
});
```

## Related Links

#2620 

## Checklist

- [ ] Tests updated (or not required).
- [ ] Documentation updated (or not required).

### CODE REVIEW

File: packages/shared/src/types/hooks.ts [hunk 1]
```diff
@@ -111,6 +119,18 @@ export type ModifyEnvironmentConfigFn = (
   utils: ModifyEnvironmentConfigUtils,
 ) => MaybePromise<MergedEnvironmentConfig | void>;
 
+export type EnvironmentContext = {
+  index: number;
+  name: string;
+  target: RsbuildTarget;
```
Comment (chenjiahan) [inline packages/shared/src/types/hooks.ts:125]: `target` is equal to `config.output.target` can we remove it.

File: packages/compat/webpack/tests/__snapshots__/webpackConfig.test.ts.snap [hunk 2]
```diff
@@ -78,7 +78,143 @@ exports[`webpackConfig > should allow to append and prepend plugins 1`] = `
     "version": 5,
   },
   HtmlBasicPlugin {
-    "environment": "web",
+    "environment": {
```
[the remark this answers is not in the document]
Comment (9aoy) [reply packages/compat/webpack/tests/__snapshots__/webpackConfig.test.ts.snap:81]: But we need pass environment info in  modifyHTMLTags hook.

[the remark this answers is not in the document]
Comment (chenjiahan) [reply packages/compat/webpack/tests/__snapshots__/webpackConfig.test.ts.snap:81]: How about using a closure

File: packages/core/src/plugins/css.ts [hunk 3]
```diff
@@ -327,9 +327,7 @@ export const pluginCss = (): RsbuildPlugin => ({
       order: 'pre',
       handler: async (chain, utils) => {
         const rule = chain.module.rule(utils.CHAIN_ID.RULE.CSS);
-        const config = api.getNormalizedConfig({
-          environment: utils.environment,
-        });
+        const config = utils.environment.config;
```
Comment (chenjiahan) [inline packages/core/src/plugins/css.ts:330]:
```suggestion
        const { config } = utils.environment;
```

File: packages/core/src/plugins/fileSize.ts [hunk 4]
```diff
@@ -192,39 +192,37 @@ export const pluginFileSize = (): RsbuildPlugin => ({
...
       if (!stats) {
         return;
       }
 
       const logs = await Promise.all(
-        Object.keys(api.context.environments).map(
-          async (environment, index) => {
-            const { printFileSize } = api.getNormalizedConfig({
+        Object.keys(environments).map(async (environment, index) => {
+          const { printFileSize } = api.getNormalizedConfig({
```
Comment (chenjiahan) [inline packages/core/src/plugins/fileSize.ts:202]: Use `environment.config` instead of `api.getNormalizedConfig`

**Review by chenjiahan — APPROVED** [review]

**Pull request merged**